python-frontend/src/main/java/org/sonar/python/tree/PythonTreeMaker.java

```diff
@@ -1676,62 +1667,93 @@ private static Expression numericLiteral(AstNode astNode) {
+    List<StringElement> elements = astNode.getChildren(PythonTokenType.STRING, PythonGrammar.FSTRING).stream()
+      .map(this::stringLiteral)
+      .filter(Objects::nonNull)
+      .collect(Collectors.toList());
     return new StringLiteralImpl(elements);
   }
 
+  private StringElementImpl stringLiteral(AstNode elementNode) {
+    Token token = toPyToken(elementNode.getToken());
+    if (token == null) {
+      return null;
+    }
+    if (elementNode.is(PythonGrammar.FSTRING)) {
+      Token fstringEnd = toPyToken(elementNode.getFirstChild(PythonTokenType.FSTRING_END).getToken());
+      List<Tree> fStringMiddles = getFStringMiddles(elementNode);
+      return new StringElementImpl(token, fStringMiddles, fstringEnd);
+    }
+    return new StringElementImpl(token, List.of(), null);
+  }
+
+  private List<Tree> getFStringMiddles(AstNode expressionNode) {
+    return expressionNode
+      .getChildren(PythonGrammar.FSTRING_REPLACEMENT_FIELD, PythonTokenType.FSTRING_MIDDLE)
+      .stream()
+      .map(fStringMiddle -> {
+        if (fStringMiddle.is(PythonGrammar.FSTRING_REPLACEMENT_FIELD)) {
+          return formattedExpression(fStringMiddle);
+        }
+        return stringLiteral(fStringMiddle);
+      })
+      .filter(Objects::nonNull)
+      .collect(Collectors.toList());
+  }
+
   private FormatSpecifier formatSpecifier(AstNode expressionNode) {
     AstNode formatSpecifierNode = expressionNode.getFirstChild(PythonGrammar.FORMAT_SPECIFIER);
     if (formatSpecifierNode == null) {
       return null;
     }
 
     Token columnToken = toPyToken(formatSpecifierNode.getFirstChild(PythonPunctuator.COLON).getToken());
-    List<FormattedExpression> nestedExpressions = formatSpecifierNode.getChildren(PythonGrammar.FORMATTED_EXPR)
-      .stream()
-      .map(this::formattedExpression)
-      .collect(Collectors.toList());
-
-    return new FormatSpecifierImpl(columnToken, nestedExpressions);
+    if (columnToken == null) {
+      return null;
+    }
+    List<Tree> fStringMiddles = getFStringMiddles(formatSpecifierNode);
+    return new FormatSpecifierImpl(columnToken, fStringMiddles);
   }
 
-  private FormattedExpression formattedExpression(AstNode expressionNode) {
-    var expressionsList = expressionNode.getFirstChild(PythonGrammar.TESTLIST);
-    var expressions = expressionsList
-      .getChildren(PythonGrammar.TEST)
-      .stream()
-      .map(this::expression)
-      .collect(Collectors.toList());
-
+  private FormattedExpressionImpl formattedExpression(AstNode expressionNode) {
     Expression exp;
-    if (expressionsList.getChildren().size() == 1) {
-      exp = expressions.get(0);
+    if (expressionNode.hasDirectChildren(PythonGrammar.YIELD_EXPR)) {
+      var yieldExpression = expressionNode.getFirstChild(PythonGrammar.YIELD_EXPR);
+      exp = expression(yieldExpression);
     } else {
+      var expressionsList = expressionNode.getFirstChild(PythonGrammar.TESTLIST_STAR_EXPR);
```

Comment:
There are already method in `PythonTreeMaker` to parse `TESTLIST_STAR_EXPR` - `org.sonar.python.tree.PythonTreeMaker#expressionList `